test/langtools/jdk/javadoc/lib/javadoc/tester/TestJavaFileManagerBuilder.java

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javadoc.tester;
+
+import javax.tools.FileObject;
+import javax.tools.JavaFileManager;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.lang.reflect.UndeclaredThrowableException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+
+/**
+ * A builder to create "test file managers" that can return "test file objects".
+ * All such objects can throw user-provided exceptions when specified methods
+ * are called. This is done by registering "handlers" to be associated with individual
+ * methods.
+ *
+ * The file objects that are returned as "test file objects" are filtered by a predicate
+ * on the file object.
+ *
+ * Note that "test file objects" passed as arguments to methods on the "test file manager"
+ * that created them are unwrapped, and replaced by the original file object.
+ * This ensures that the underlying file manager sees the underlying file objects,
+ * for cases when the identity of the file objects is important.
+ * However, it does mean that methods on file objects called internally by a
+ * file manager will not throw any user-provided exceptions.
+ *
+ * For now, the handlers for a file object are simply grouped by predicate and then by
+ * method, and the group of methods used for a "test file object" is determined by the
+ * first predicate that matches.
+ * An alternative, more expensive, implementation would be to group the handlers
+ * by method and predicate and then dynamically build the set of methods to be used for
+ * a file object by filtering the methods by their applicable predicate.
+ */
```

Comment:
> [snip]. Update the copyright years before integrating.

Done

Reply:
> These 1.5 paragraphs seem no longer relevant.

I think the comments are still useful/relevant because the most recent change was just about avoid the use of a predicate as a map key, but the ordered list off predicates is still there, albeit in a slightly different form.